frame/evm/src/runner/stack.rs

```diff
@@ -663,17 +739,56 @@ impl<'vicinity, 'config, T: Config> BackendT for SubstrateStackState<'vicinity,
 		}
 	}
 
-	fn code(&self, address: H160) -> Vec<u8> {
-		<AccountCodes<T>>::get(address)
+	#[cfg(feature = "evm-with-weight-limit")]
+	fn code(&mut self, address: H160) -> Result<Vec<u8>, ExitError> {
+		let maybe_record = !<AccountCodesAccessed<T>>::get(address);
+		// Skip if the address has been already recorded this block
```

Comment:
Hmmm, so we only record extra weight if it hasn't been read in a block? That means that some txns will succeed only if some other txn warmed up its state reads before it.

The per-block approach is nice for efficiency (it will prevent this "new feature" from reducing block fullness), but I don't like that it makes txn outcomes depend on other txns in the same block.

It could also do weird things like incentivize a project (think a popular DEX) to send do-nothing txns that serve only to warm up reads for their users' txns.

Also, if we went with a per-txn solution instead of per-block, we could probably avoid the extra storage overhead of <AccountCodesAccessed> (right?)

Reply:
> I don't like that it makes txn outcomes depend on other txns in the same block

I guess that's the trade-off vs. over accounting, it's up for debate what we do prefer of course. 

> It could also do weird things like incentivize a project (think a popular DEX) to send do-nothing txns that serve only to warm up reads for their users' txns.

why would that be an issue pov-size-wise?

Reply:
> why would that be an issue pov-size-wise?

It would just waste block space. They'd pay for it, so not a big deal I guess. But the incentive seems perverse.

> I guess that's the trade-off vs. over accounting, it's up for debate what we do prefer of course.

An alternative / hybrid might work: keep the per-block accounting for efficiency (allow maximum block fullness), but still charge each transaction on a per-txn basis (make the txn outcome independent of previous reads in the block).

Agree that it's a trade off, though...

Reply:
> The only difference is that we will not artifically OutOfGas transactions based on PovSize that was already part of the merkle proof.

This isn't how other things work -- in Substrate we benchmark things to see the *worst case* PoV impact and charge based on that. This means that transactions aren't dependent (in terms of PoV) on previous transactions or existing proof.

Similarly, the worm/cold storage system in Ethereum works on a per-txn basis, not a per-block one.

Reply:
For transparency: @tgmichel and I discussed this a bit more and agreed that this PR in its current form is a good initial implementation and that some future work could address the issue described here.